src/formats/SMI.cpp

```diff
@@ -0,0 +1,705 @@
+            continue;
+        }
+
+        // We've seen this neighbor before! Ring found
+        if (hit_atoms[neighbor]) {
+            // Don't process the same ring connection twice
+            if (ring_bonds.count(Bond(neighbor, current_atom))) {
+                continue;
+            }
+            ring_bonds.insert(Bond(neighbor, current_atom));
+
+            auto neigh_ring_iter = ring_atoms.find(neighbor);
+            if (neigh_ring_iter == ring_atoms.end()) {
+                neigh_ring_iter = ring_atoms.insert({neighbor, 0}).first;
+            }
+            ++neigh_ring_iter->second;
+
+            continue; // No need to see this atom again
+        }
+
+        // Continue the search
+        find_rings_helper(adj_list, hit_atoms, ring_bonds,
+                          ring_atoms, neighbor, current_atom);
+    }
+}
+
+static std::map<size_t, size_t> find_rings(
+    const std::vector<std::vector<size_t>>& adj_list) {
+
+    std::map<size_t, size_t> ring_atoms;
+    std::vector<bool> hit_atoms(adj_list.size(), false);
+    std::set<Bond> ring_bonds;
+    
+    while (!all(hit_atoms)) {
+        auto not_hit = std::find(hit_atoms.begin(), hit_atoms.end(), false);
+        auto current_atom = static_cast<size_t>(std::distance(hit_atoms.begin(), not_hit));
+        auto& current_atom_bonds = adj_list[current_atom];
+        hit_atoms[current_atom] = true;
+        if (current_atom_bonds.empty()) {
+            continue;
+        }
+
+        find_rings_helper(adj_list, hit_atoms, ring_bonds, ring_atoms,
+                          current_atom_bonds[0], current_atom
+        );
+    }
+
+    return ring_atoms;
+}
+
+static void write_atom_smiles(std::unique_ptr<TextFile>& file_, const Atom& atom) {
+    bool needs_brackets = false;
+    auto type = atom.type();
+
+    // The mass must be an integer is the only check we need as all atoms in the
+    // periodic table do have non-integer masses. Therefore, if the the mass is
+    // not an integer, then we know the user has set an isotope
+    double mass;
+    size_t mass_int = 0;
+    if (std::modf(atom.mass(), &mass) == 0.0 && mass != 0.0) {
+        needs_brackets = true;
+        mass_int = static_cast<size_t>(mass);
+    }
+
+    // If any of these values are set / not a default value
+    auto smi_class = atom.get("smiles_class");
+    auto chirality = atom.get<Property::STRING>("chirality").value_or("");
+    auto explicit_h =atom.get<Property::DOUBLE>("hydrogen_count").value_or(0.0);
+
+    // Charge is similar to mass. It must be an integer, otherwise it is ignored
+    double charge;
+    int charge_int = 0;
+    if (std::modf(atom.charge(), &charge) == 0.0 && charge != 0.0) {
+        needs_brackets = true;
+        charge_int = static_cast<int>(charge);
+    }
+
+    // Before the atom is printed, we must know if we are printing a property set.
+    if (explicit_h != 0.0 || smi_class || !chirality.empty() ||
+        aliphatic_organic.count(type) == 0) {
```

Comment:
On the readability side, could the condition go on the above line? Or add some separation with the next line?